### PR TITLE
Feature/gpumap

### PR DESCRIPTION
**Pull request by Ben-Epstein**
supports applying umap to embeddings and data embeddings on the local client. 

Data embeddings are supported for train/test/val/inference

Regular embeddings are only supported for train/test/val. Inference will be a next step, and I'll need some support from @elboy3, but for now customers only need this for train/test/val, so we are fine to proceed.

Aside from the code changes, a few functions (`_shutil_rmtree_retry`, `concat_hdf5_files`, `save_hdf5_file`) were in the wrong modules, so I moved them over (that's why youll see large functions removed). No code was changed, just moved them into the proper modules.

We pin our cuda libs because nvidia recently broke everything by moving to a new hosted environment for packages https://github.com/rapidsai/cudf/issues/12762#issuecomment-1462845289

## Tested

I tested this e2e with many many runs. See in dev, all runs in the project `gpu_umap` 
* [emotion](https://console.dev.rungalileo.io/insights?projectId=8c145e49-4328-4820-b128-08839abeed94&runId=a9baf860-014d-4423-a185-27070b644c55)
* [newsgroups](https://console.dev.rungalileo.io/insights?projectId=8c145e49-4328-4820-b128-08839abeed94&runId=b93444ec-90fd-435e-9768-2174c22aee48)
* [conv_intent](https://console.dev.rungalileo.io/insights?projectId=8c145e49-4328-4820-b128-08839abeed94&runId=a69aff23-4b0c-490b-9b53-554e06ec335b&tab=em)
* [mit_movies](https://console.dev.rungalileo.io/insights?projectId=8c145e49-4328-4820-b128-08839abeed94&runId=278c9f30-feda-4240-97f8-87a76282ba90&tab=em&taskType=2)
* [wnut_17](https://console.dev.rungalileo.io/insights?projectId=8c145e49-4328-4820-b128-08839abeed94&runId=dc82bbdd-9cc4-4df4-b009-e6872f9c44d7&taskType=2)

And a few others... Check dev for all the runs in that project.

### CODE REVIEW

File: dataquality/loggers/data_logger/text_classification.py [hunk 1]
```diff
@@ -477,8 +477,9 @@ def separate_dataframe(
             emb_cols = ["id"]
             other_cols = ["id"]
         else:
-            emb_cols = ["id", "emb"]
-            ignore_cols = ["emb", "split_id"] + prob_cols
+            emb_cols = ["id", "emb", "x", "y", "emb_pca"]
```
Comment (elboy3) [inline dataquality/loggers/data_logger/text_classification.py:480]: are the cols data_x and data_y for data embs?

Comment (Ben-Epstein) [reply dataquality/loggers/data_logger/text_classification.py:480]: yup thats how we save it in the final df in minio. but those are updated immediately so aren't affected here

**Review by elboy3 — COMMENTED** [review]
One q about data_x and data_y, then ✅

File: dataquality/utils/cuda.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+        return True
+    except ImportError:
+        return False
+
+
+def get_pca_embeddings(embs: np.ndarray) -> np.ndarray:
+    """Uses Cuda and GPUs to create the PCA embeddings before uploading
+
+    Should only be called if cuda ML available (`cuda_available()`)
```
Comment (anthonycorletti) [inline dataquality/utils/cuda.py:16]: wdyt about warning if cuml is not available and this is called?

File: dataquality/utils/cuda.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    """
+    import cuml
+
+    pca = cuml.PCA(n_components=100)
```
Comment (anthonycorletti) [inline dataquality/utils/cuda.py:20]: what makes 100 the special number?

File: dataquality/utils/cuda.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+
+def get_umap_embeddings(embs: np.ndarray) -> np.ndarray:
+    """Uses Cuda and GPUs to create the UMAP embeddings before uploading
+
+    Should only be called if cuda ML available (`cuda_available()`)
+    """
+    import cuml
+
+    umap = cuml.UMAP(n_neighbors=15, n_components=2, min_dist=0.25)
```
Comment (anthonycorletti) [inline dataquality/utils/cuda.py:31]: should these params be adjustable?

File: pyproject.toml [hunk 5]
```diff
@@ -77,5 +77,16 @@ dev = [
     "jupyter==1.0.0",
 ]
 
+# Need to be installed with `pip install 'dataquality[cuda]' --extra-index-url=https://pypi.ngc.nvidia.com/ `
+cuda = [
+    "ucx-py-cu11<0.30",
```
Comment (anthonycorletti) [inline pyproject.toml:82]: what's causing the pin under 0.30?

File: scripts/install.sh [hunk 6]
```diff
@@ -3,4 +3,4 @@
 pip install --upgrade pip
 pip install flit
 
-flit install --deps=all --extras=all --symlink
+flit install --deps=develop --symlink
```
Comment (anthonycorletti) [inline scripts/install.sh:6]: does this install test too?

File: dataquality/utils/cuda.py (hunk 4 above)
Comment (Ben-Epstein) [reply dataquality/utils/cuda.py:31]: nope - its what we do in the server, so keeping consistent

File: pyproject.toml (hunk 5 above)
Comment (Ben-Epstein) [reply pyproject.toml:82]: nvidia moved where they hosted these libraries recently, and it broke everything: https://github.com/rapidsai/cudf/issues/12762#issuecomment-1462845289

File: scripts/install.sh (hunk 6 above)
Comment (Ben-Epstein) [reply scripts/install.sh:6]: yup! test, doc, dev https://flit.pypa.io/en/stable/cmdline.html#cmdoption-flit-install-deps

**Review by elboy3 — APPROVED** [review]
🚀

**Pull request merged**